src/plugins/_manager.ts

```diff
@@ -0,0 +1,171 @@
+      // if needed, we can make the loader smarter and enable this, but not for now.
+      throw new Error(`Unsupported package reference: ${pkg}. Can either be an NPM package name, or a local path to a directory`);
+    }
+
+    const local = path.isAbsolute(pkg)
+      // relative paths must start with '.' because otherise they
+      // are indistinguishable from npm package names.
+      || pkg.startsWith(`.${path.sep}`);
+
+    // local plugins are loaded directly, npm packages are loaded
+    // from the plugins directory.
+    const modulePath = local ? path.resolve(process.cwd(), pkg) : this.pluginDir(pkg, version);
+
+    try {
+      return this.require(modulePath, version);
+    } catch (e: any) {
+
+      if (![MODULE_NOT_FOUND_ERROR_CODE, MODULE_VERSION_MISMATCH_ERROR_CODE].includes(e.code)) {
+        // some unexpected error
+        throw e;
+      }
+
+      if (local) {
+        // if a local plugin is missing, nothing we can do about it
+        throw e;
+      }
+
+      // otherwise, we install from npm and re-require.
+      this.installPackage(pkg, version, installEnv);
+      return this.require(modulePath, version);
+    }
+
+  }
+
+  private installPackage(pkg: string, version: string, env: { [key: string]: any }) {
+
+    const pluginDir = this.pluginDir(pkg, version);
+    fs.mkdirpSync(pluginDir);
+
+    const command = [
+      'npm',
+      'install', `${pkg}@${version}`,
+      '--no-save',
+      '--prefix', pluginDir,
+    ].join(' ');
+
+    const finalEnv = { ...process.env };
+    for (const [key, value] of Object.entries(env)) {
+      finalEnv[key] = typeof value === 'string' ? key : JSON.stringify(value);
+    }
+    child.execSync(command, { stdio: ['ignore', 'pipe', 'pipe'], env: finalEnv });
+  }
+
+  private require(pkg: string, version: string): Package {
+
+    const modulePath = require.resolve(pkg);
+
+    // eslint-disable-next-line @typescript-eslint/no-require-imports
+    const module = require(modulePath);
+
+    const manifestPath = this.findPackageJson(modulePath, []);
+    const manifest = fs.readJsonSync(manifestPath);
+    if (manifest.version !== version) {
+      throw new ModuleVersionMismatchError(pkg, manifest.version, version);
+    }
+
+    return { version: manifest.version, pkg: manifest.name, path: modulePath, module };
+  }
+
+  private pluginDir(pkg: string, version: string) {
+    return path.join(this.dir, pkg, version, 'node_modules', pkg);
+  }
+
+  private findPackageJson(fdp: string, searched: string[]): string {
+
+    const packageJsonPath = path.join(fdp, 'package.json');
+    if (fs.existsSync(packageJsonPath)) {
+      return packageJsonPath;
+    }
+
+    if (path.dirname(fdp) === fdp) {
+      throw new Error(`Unable to locate package.json file. Searched in: ${searched.join(os.EOL)}`);
+    }
+
+    return this.findPackageJson(path.dirname(fdp), [...searched, packageJsonPath]);
+  }
```

Comment:
The hot path of this code is for checking our own plugins directory, which we control so we know the structure of. 
Its true that this is also called for local plugins, and may fail in some environments. Im kind of ok with handling those as they come, especially since it was designed to make plugin **development** easier, not **consumption**. 

Now that I think of it though, the local case creates too much diversion in the code. I'm leaning towards getting rid of it altogether, and have plugin developers symlink their plugin into the expected plugins directory, in the expected structure. What do you think?

Reply:
As it turns out, this whole thing is unnecessary because we already encode the plugin version in the path.